grakel_replace/torch_wl_kernel.py

```diff
@@ -1,13 +1,122 @@
+        n_iter: int = 5,
+        *,
+        normalize: bool = True,
+        active_dims: tuple[int, ...],
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(active_dims=active_dims, **kwargs)
+        self.graph_lookup = graph_lookup
+        self.n_iter = n_iter
+        self.normalize = normalize
+
+        # NOTE: set in the `super().__init__()`
+        self.active_dims: torch.Tensor
+
+    def set_graph_lookup(self, graph_lookup: list[nx.Graph]) -> None:
+        self.graph_lookup = graph_lookup
+
+    def forward(
+        self,
+        x1: torch.Tensor,
+        x2: torch.Tensor,
+        *,
+        diag: bool = False,
+        last_dim_is_batch: bool = False,
+        **params: Any,
+    ):
+        if last_dim_is_batch:
+            raise NotImplementedError("TODO: Figure this out")
+
+        assert x1.shape[-1] == 1, "Last dimension must be the graph index"
+        assert x2.shape[-1] == 1, "Last dimension must be the graph index"
+
+        # TODO: Optimizations
+        #
+        # 1. We're computing the whole K Matrix, but we only need the K_x1_x2
+        #
+        #           K
+        #       --------------------
+        #       | K_x1_x1  K_x1_x2 |
+        #       | K_x2_x1  K_x2_x2 |
+        #       --------------------
+        #
+        # However in the case where x1 == x2, we can shortcut this slightly as in
+        # the above, K_x1_x2 == K_x2_x1 == K_x1_x1 == K_x2_x2
+        # This shortcut is implemented below based on this flag.
+        #
+        # 2. The _TorchWLKernel used below has the following properties, which
+        #   get set on forward. In the case where x1.ndim == 3 then the first dim is
+        #   the `q` dim. Doesn't matter what it is other than we end up repeating the
+        #   processing the graphs `q` times. Given that it's likely that the indices
+        #   in last dimension are likely to be constant (i.e. all `4`, indicating the
+        #   `4th` graph, we are effectively doing a lot of extra calculation. We could
+        #   shortcut this by pre-computing these for each index. Could be nice to somehow
+        #   have the inned `_TorchWLKernel` be aware of this extra dimension but it's
+        #   fine if not as long as we can reduce the extraneuous computations. We could
+        #   change the interface of `_TorchWLKernel` to take in the raw processed
+        #   tensors instead of `nx.Graph` objects, which we would instead preprocess here.
+        #   If that's the case, we could move the `_TorchWLKernel` to essentially just
+        #   be functions we call instead with the correct pre-processed data.
+        #
+        #       .self.label_dict
+        #       .self.label_counter
+        #
+        x1_is_x2 = torch.equal(x1, x2)
+
+        # NOTE: The active dim is already selected out for us and is the last dimension
+        # (not including whatever happens when last_dim_is_batch) is True.
+        if x1.ndim == 3:
```

Comment:
This is to do with the batching that goes on during `optimize_acqf`.

It corresponds with the `raw_samples=` parameter